AssertionError on MendelErrors

```
Traceback (most recent call last):
  File "/tmp/837e8bf7-96ab-442b-8777-e5f05e247b63/generate_qc_annotations.py", line 223, in <module>
    try_slack(args.slack_channel, main, args)
  File "/tmp/837e8bf7-96ab-442b-8777-e5f05e247b63/pyscripts_KndB0n.zip/gnomad_hail/slack_utils.py", line 97, in try_slack
  File "/tmp/837e8bf7-96ab-442b-8777-e5f05e247b63/pyscripts_KndB0n.zip/gnomad_hail/slack_utils.py", line 77, in try_slack
  File "/tmp/837e8bf7-96ab-442b-8777-e5f05e247b63/generate_qc_annotations.py", line 194, in main
    vds, sample_table = generate_family_stats(vds, fam_file)
  File "/tmp/837e8bf7-96ab-442b-8777-e5f05e247b63/generate_qc_annotations.py", line 134, in generate_family_stats
    _, _, adj_per_sample, adj_per_variant = hl.mendel_errors(vds, ped)
  File "<decorator-gen-616>", line 2, in mendel_errors
  File "/tmp/837e8bf7-96ab-442b-8777-e5f05e247b63/hail-devel-48d0534b1b09.zip/hail/typecheck/check.py", line 489, in _typecheck
  File "/tmp/837e8bf7-96ab-442b-8777-e5f05e247b63/hail-devel-48d0534b1b09.zip/hail/methods/family_methods.py", line 205, in mendel_errors
  File "/usr/lib/spark/python/lib/py4j-0.10.4-src.zip/py4j/java_gateway.py", line 1133, in __call__
  File "/tmp/837e8bf7-96ab-442b-8777-e5f05e247b63/hail-devel-48d0534b1b09.zip/hail/utils/java.py", line 238, in deco
hail.utils.java.FatalError: AssertionError: assertion failed

Java stack trace:
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:156)
	at is.hail.rvd.OrderedRVDPartitioner.getPartitionPK(OrderedRVDPartitioner.scala:59)
	at is.hail.sparkextras.OrderedDependency$.getDependencies(OrderedRDD2.scala:22)
	at is.hail.sparkextras.OrderedJoinDistinctRDD2$$anonfun$getPartitions$1.apply(OrderedRDD2.scala:42)
	at is.hail.sparkextras.OrderedJoinDistinctRDD2$$anonfun$getPartitions$1.apply(OrderedRDD2.scala:39)
	at scala.Array$.tabulate(Array.scala:331)
	at is.hail.sparkextras.OrderedJoinDistinctRDD2.getPartitions(OrderedRDD2.scala:39)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.RDD.getNumPartitions(RDD.scala:266)
	at is.hail.rvd.RVD$class.getNumPartitions(RVD.scala:128)
	at is.hail.rvd.OrderedRVD.getNumPartitions(OrderedRVD.scala:19)
	at is.hail.sparkextras.OrderedJoinDistinctRDD2.getPartitions(OrderedRDD2.scala:39)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.UnionRDD$$anonfun$1.apply(UnionRDD.scala:84)
	at org.apache.spark.rdd.UnionRDD$$anonfun$1.apply(UnionRDD.scala:84)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.immutable.List.map(List.scala:285)
	at org.apache.spark.rdd.UnionRDD.getPartitions(UnionRDD.scala:84)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.Partitioner$$anonfun$defaultPartitioner$2.apply(Partitioner.scala:66)
	at org.apache.spark.Partitioner$$anonfun$defaultPartitioner$2.apply(Partitioner.scala:66)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.immutable.List.map(List.scala:285)
	at org.apache.spark.Partitioner$.defaultPartitioner(Partitioner.scala:66)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$reduceByKey$3.apply(PairRDDFunctions.scala:329)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$reduceByKey$3.apply(PairRDDFunctions.scala:329)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:362)
	at org.apache.spark.rdd.PairRDDFunctions.reduceByKey(PairRDDFunctions.scala:328)
	at is.hail.methods.MendelErrors.nErrorPerNuclearFamily(MendelErrors.scala:144)
	at is.hail.methods.MendelErrors.fMendelKT(MendelErrors.scala:183)
	at is.hail.variant.MatrixTable.mendelErrors(MatrixTable.scala:1915)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:748)

Hail version: devel-42db165
Error summary: AssertionError: assertion failed
```
Pipeline (happens on the 2nd mendel_errrors):
```
    vds = vds.select_rows(vds.locus, vds.alleles)
    fam_kt = hl.import_fam(fam_file)
    vds = vds.annotate_cols(fam=fam_kt[vds.s])

    # Unphased for now, since mendel_errors does not support phased alleles
    vds = vds.annotate_entries(GT=hl.call(vds.GT[0], vds.GT[1], phased=False))
    ped = hl.Pedigree.read(fam_file)
    tdt_table = hl.tdt(vds, ped)
    _, _, per_sample, per_variant = hl.mendel_errors(vds, ped)
    family_stats = Struct(mendel=per_variant[(vds.locus, vds.alleles)],
                          tdt=tdt_table[(vds.locus, vds.alleles)],
                          ac_unrelated_qc=hl.agg.sum(hl.agg.filter(
                              True & hl.is_missing(vds.fam.pat_id),
                              vds.GT.num_alt_alleles())),
                          meta={'group': 'raw'})
    vds = vds.annotate_rows(family_stats=[family_stats])

    vds = filter_to_adj(vds)
    adj_tdt_table = hl.tdt(vds, ped)
    _, _, adj_per_sample, adj_per_variant = hl.mendel_errors(vds, ped)
    family_stats_adj = Struct(mendel=adj_per_variant[(vds.locus, vds.alleles)],
                              tdt=adj_tdt_table[(vds.locus, vds.alleles)],
                              ac_unrelated_qc=hl.agg.sum(hl.agg.filter(
                                  True & hl.is_missing(vds.fam.pat_id),
                                  vds.GT.num_alt_alleles())),
                              meta={'group': 'adj'})

    per_sample = per_sample.annotate(adj=adj_per_sample[per_sample.s])

    vds = vds.annotate_rows(family_stats=vds.family_stats.append(family_stats_adj))
```

--- a/src/main/scala/is/hail/rvd/OrderedRVD.scala
+++ b/src/main/scala/is/hail/rvd/OrderedRVD.scala
@@ -114,23 +114,21 @@ class OrderedRVD private(
 
   def orderedZipJoin(right: OrderedRVD): OrderedZipJoinRDD = {
     val pkOrd = this.partitioner.pkType.ordering
-    if (pkOrd.gt(this.partitioner.minBound, right.partitioner.minBound) ||
-      pkOrd.lt(this.partitioner.maxBound, right.partitioner.maxBound)) {
-
+    if (this.partitioner.range.includes(pkOrd, right.partitioner.range))
+      new OrderedZipJoinRDD(this, right)
+    else {
       val newRangeBounds = partitioner.rangeBounds.toArray
-
+      val newStart = pkOrd.min(this.partitioner.range.start, right.partitioner.range.start)
+      val newEnd = pkOrd.max(this.partitioner.range.end, right.partitioner.range.end)
       newRangeBounds(0) = newRangeBounds(0).asInstanceOf[Interval]
-        .copy(start = pkOrd.min(this.partitioner.minBound, right.partitioner.minBound))
+        .copy(start = newStart, includeStart = true)
       newRangeBounds(newRangeBounds.length - 1) = newRangeBounds(newRangeBounds.length - 1).asInstanceOf[Interval]
-        .copy(end = pkOrd.max(this.partitioner.maxBound, right.partitioner.maxBound))
+        .copy(end = newEnd, includeEnd = true)
 
       val newPartitioner = new OrderedRVDPartitioner(partitioner.partitionKey,
         partitioner.kType, UnsafeIndexedSeq(partitioner.rangeBoundsType, newRangeBounds))
 
       new OrderedZipJoinRDD(OrderedRVD(this.typ, newPartitioner, this.rdd), right)
-
-    } else {
-      new OrderedZipJoinRDD(this, right)
     }
   }
 

--- a/src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala
+++ b/src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala
@@ -45,9 +45,6 @@ class OrderedRVDPartitioner(
 
   def range: Option[Interval] = rangeTree.root.map(_.range)
 
-  def minBound: Any = rangeBounds(0).asInstanceOf[Interval].start
-  def maxBound: Any = rangeBounds(numPartitions - 1).asInstanceOf[Interval].end
-
   // if outside bounds, return min or max depending on location
   // pk: Annotation[pkType]
   def getPartitionPK(pk: Any): Int = {
@@ -79,10 +76,10 @@ class OrderedRVDPartitioner(
 
     part match {
       case Array() =>
-        if (pkType.ordering.lt(pkUR, minBound))
+        if (range.forall(_.greaterThanPoint(pkType.ordering, pkUR)))
           0
         else {
-          assert(pkType.ordering.gt(pkUR, maxBound))
+          assert(range.forall(_.lessThanPoint(pkType.ordering, pkUR)))
           numPartitions - 1
         }
       case Array(x) => x

--- a/src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala
+++ b/src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala
@@ -43,7 +43,7 @@ class OrderedRVDPartitioner(
 
   def loadEnd(i: Int): Long = pkIntervalType.loadStart(region, loadElement(i))
 
-  def range: Option[Interval] = rangeTree.root.map(_.range)
+  def range: Interval = rangeTree.root.get.range
 
   // if outside bounds, return min or max depending on location
   // pk: Annotation[pkType]
@@ -52,10 +52,10 @@ class OrderedRVDPartitioner(
     val part = rangeTree.queryValues(pkType.ordering, pk)
     part match {
       case Array() =>
-        if (range.forall(_.greaterThanPoint(pkType.ordering, pk)))
+        if (range.greaterThanPoint(pkType.ordering, pk))
           0
         else {
-          assert(range.forall(_.lessThanPoint(pkType.ordering, pk)))
+          assert(range.lessThanPoint(pkType.ordering, pk))
           numPartitions - 1
         }
 
@@ -76,10 +76,10 @@ class OrderedRVDPartitioner(
 
     part match {
       case Array() =>
-        if (range.forall(_.greaterThanPoint(pkType.ordering, pkUR)))
+        if (range.greaterThanPoint(pkType.ordering, pkUR))
           0
         else {
-          assert(range.forall(_.lessThanPoint(pkType.ordering, pkUR)))
+          assert(range.lessThanPoint(pkType.ordering, pkUR))
           numPartitions - 1
         }
       case Array(x) => x

--- a/src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala
+++ b/src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala
@@ -52,10 +52,10 @@ class OrderedRVDPartitioner(
     val part = rangeTree.queryValues(pkType.ordering, pk)
     part match {
       case Array() =>
-        if (range.greaterThanPoint(pkType.ordering, pk))
+        if (range.isAbovePosition(pkType.ordering, pk))
           0
         else {
-          assert(range.lessThanPoint(pkType.ordering, pk))
+          assert(range.isBelowPosition(pkType.ordering, pk))
           numPartitions - 1
         }
 
@@ -76,10 +76,10 @@ class OrderedRVDPartitioner(
 
     part match {
       case Array() =>
-        if (range.greaterThanPoint(pkType.ordering, pkUR))
+        if (range.isAbovePosition(pkType.ordering, pkUR))
           0
         else {
-          assert(range.lessThanPoint(pkType.ordering, pkUR))
+          assert(range.isBelowPosition(pkType.ordering, pkUR))
           numPartitions - 1
         }
       case Array(x) => x

--- a/src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala
+++ b/src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala
@@ -43,6 +43,8 @@ class OrderedRVDPartitioner(
 
   def loadEnd(i: Int): Long = pkIntervalType.loadStart(region, loadElement(i))
 
+  def range: Option[Interval] = rangeTree.root.map(_.range)
+
   def minBound: Any = rangeBounds(0).asInstanceOf[Interval].start
   def maxBound: Any = rangeBounds(numPartitions - 1).asInstanceOf[Interval].end
 
@@ -53,10 +55,10 @@ class OrderedRVDPartitioner(
     val part = rangeTree.queryValues(pkType.ordering, pk)
     part match {
       case Array() =>
-        if (pkType.ordering.lt(pk, minBound))
+        if (range.forall(_.greaterThanPoint(pkType.ordering, pk)))
           0
         else {
-          assert(pkType.ordering.gt(pk, maxBound))
+          assert(range.forall(_.lessThanPoint(pkType.ordering, pk)))
           numPartitions - 1
         }
 

--- a/src/main/scala/is/hail/sparkextras/OrderedRDD2.scala
+++ b/src/main/scala/is/hail/sparkextras/OrderedRDD2.scala
@@ -75,8 +75,7 @@ class OrderedZipJoinRDD(left: OrderedRVD, right: OrderedRVD)
     Seq[Dependency[_]](new OneToOneDependency(left.rdd),
       new OrderedDependency(left, right))) {
 
-  assert(left.partitioner.pkType.ordering.lteq(left.partitioner.minBound, right.partitioner.minBound) &&
-    left.partitioner.pkType.ordering.gteq(left.partitioner.maxBound, right.partitioner.maxBound))
+  assert(left.partitioner.range.includes(left.partitioner.pkType.ordering, right.partitioner.range))
 
   private val leftPartitionForRightRow = new OrderedRVDPartitioner(
     right.typ.partitionKey,

--- a/src/main/scala/is/hail/utils/IntervalTree.scala
+++ b/src/main/scala/is/hail/utils/IntervalTree.scala
@@ -21,6 +21,15 @@ case class Interval(start: Any, end: Any, includeStart: Boolean, includeEnd: Boo
       false
   }
 
+  def includes(pord: ExtendedOrdering, other: Interval): Boolean =
+  other.definitelyEmpty(pord) || ({
+    val cstart = pord.compare(this.start, other.start)
+    cstart < 0 || ((!other.includeStart || this.includeStart) && cstart == 0)
+  } && {
+    val cend = pord.compare(this.end, other.end)
+    cend > 0 || ((!other.includeEnd || this.includeEnd) && cend == 0)
+  })
+
   def mayOverlap(pord: ExtendedOrdering, other: Interval): Boolean = {
     !definitelyDisjoint(pord, other)
   }

--- a/src/main/scala/is/hail/utils/IntervalTree.scala
+++ b/src/main/scala/is/hail/utils/IntervalTree.scala
@@ -22,25 +22,25 @@ case class Interval(start: Any, end: Any, includeStart: Boolean, includeEnd: Boo
   }
 
   def includes(pord: ExtendedOrdering, other: Interval): Boolean =
-  other.definitelyEmpty(pord) || ({
-    val cstart = pord.compare(this.start, other.start)
-    cstart < 0 || ((!other.includeStart || this.includeStart) && cstart == 0)
-  } && {
-    val cend = pord.compare(this.end, other.end)
-    cend > 0 || ((!other.includeEnd || this.includeEnd) && cend == 0)
-  })
+    other.definitelyEmpty(pord) || ({
+      val cstart = pord.compare(this.start, other.start)
+      cstart < 0 || (cstart == 0 && (!other.includeStart || this.includeStart))
+    } && {
+      val cend = pord.compare(this.end, other.end)
+      cend > 0 || (cend == 0 && (!other.includeEnd || this.includeEnd))
+    })
 
   def mayOverlap(pord: ExtendedOrdering, other: Interval): Boolean = {
     !definitelyDisjoint(pord, other)
   }
 
-  def greaterThanPoint(pord: ExtendedOrdering, p: Any): Boolean =
+  def isAbovePosition(pord: ExtendedOrdering, p: Any): Boolean =
     definitelyEmpty(pord) || {
       val c = pord.compare(p, start)
       c < 0 || (!includeStart && c == 0)
     }
 
-  def lessThanPoint(pord: ExtendedOrdering, p: Any): Boolean =
+  def isBelowPosition(pord: ExtendedOrdering, p: Any): Boolean =
     definitelyEmpty(pord) || {
       val c = pord.compare(p, end)
       c > 0 || (!includeEnd && c == 0)
@@ -48,7 +48,7 @@ case class Interval(start: Any, end: Any, includeStart: Boolean, includeEnd: Boo
 
   def definitelyDisjoint(pord: ExtendedOrdering, other: Interval): Boolean =
     definitelyEmpty(pord) || other.definitelyEmpty(pord) ||
-      disjointAndLessThan(pord, other) || disjointAndGreaterThan(pord, other)
+      isBelow(pord, other) || isAbove(pord, other)
 
   private var _emptyDefined: Boolean = false
   private var _empty: Boolean = false
@@ -70,19 +70,19 @@ case class Interval(start: Any, end: Any, includeStart: Boolean, includeEnd: Boo
       "includeStart" -> TBoolean().toJSON(includeStart),
       "includeEnd" -> TBoolean().toJSON(includeEnd))
 
-  def disjointAndLessThan(pord: ExtendedOrdering, other: Interval): Boolean =
+  def isBelow(pord: ExtendedOrdering, other: Interval): Boolean =
     this.definitelyEmpty(pord) || other.definitelyEmpty(pord) || {
       val c = pord.compare(this.end, other.start)
       c < 0 || (c == 0 && (!this.includeEnd || !other.includeStart))
     }
 
-  def disjointAndGreaterThan(pord: ExtendedOrdering, other: Interval): Boolean =
+  def isAbove(pord: ExtendedOrdering, other: Interval): Boolean =
     this.definitelyEmpty(pord) || other.definitelyEmpty(pord) || {
       val c = pord.compare(this.start, other.end)
       c > 0 || (c == 0 && (!this.includeStart || !other.includeEnd))
     }
 
-  def mergeable(pord: ExtendedOrdering, other: Interval): Boolean =
+  def canMergeWith(pord: ExtendedOrdering, other: Interval): Boolean =
     this.definitelyEmpty(pord) || other.definitelyEmpty(pord) || ({
       val c = pord.compare(this.start, other.end)
       c < 0 || (c == 0 && (this.includeStart || other.includeEnd))
@@ -92,7 +92,7 @@ case class Interval(start: Any, end: Any, includeStart: Boolean, includeEnd: Boo
     })
 
   def merge(pord: ExtendedOrdering, other: Interval): Option[Interval] = {
-    if (mergeable(pord, other)) {
+    if (canMergeWith(pord, other)) {
       if (other.definitelyEmpty(pord))
         Some(this)
       else if (this.definitelyEmpty(pord))

--- a/src/main/scala/is/hail/utils/IntervalTree.scala
+++ b/src/main/scala/is/hail/utils/IntervalTree.scala
@@ -25,6 +25,18 @@ case class Interval(start: Any, end: Any, includeStart: Boolean, includeEnd: Boo
     !definitelyDisjoint(pord, other)
   }
 
+  def greaterThanPoint(pord: ExtendedOrdering, p: Any): Boolean =
+    definitelyEmpty(pord) || {
+      val c = pord.compare(p, start)
+      c < 0 || (!includeStart && c == 0)
+    }
+
+  def lessThanPoint(pord: ExtendedOrdering, p: Any): Boolean =
+    definitelyEmpty(pord) || {
+      val c = pord.compare(p, end)
+      c > 0 || (!includeEnd && c == 0)
+    }
+
   def definitelyDisjoint(pord: ExtendedOrdering, other: Interval): Boolean =
     definitelyEmpty(pord) || other.definitelyEmpty(pord) ||
       disjointAndLessThan(pord, other) || disjointAndGreaterThan(pord, other)

--- a/src/main/scala/is/hail/utils/IntervalTree.scala
+++ b/src/main/scala/is/hail/utils/IntervalTree.scala
@@ -37,13 +37,13 @@ case class Interval(start: Any, end: Any, includeStart: Boolean, includeEnd: Boo
   def isAbovePosition(pord: ExtendedOrdering, p: Any): Boolean =
     definitelyEmpty(pord) || {
       val c = pord.compare(p, start)
-      c < 0 || (!includeStart && c == 0)
+      c < 0 || (c == 0 && !includeStart)
     }
 
   def isBelowPosition(pord: ExtendedOrdering, p: Any): Boolean =
     definitelyEmpty(pord) || {
       val c = pord.compare(p, end)
-      c > 0 || (!includeEnd && c == 0)
+      c > 0 || (c == 0 && !includeEnd)
     }
 
   def definitelyDisjoint(pord: ExtendedOrdering, other: Interval): Boolean =

--- a/src/test/scala/is/hail/utils/IntervalSuite.scala
+++ b/src/test/scala/is/hail/utils/IntervalSuite.scala
@@ -60,6 +60,14 @@ class IntervalSuite extends TestNGSuite {
     }
   }
 
+  @Test def interval_agrees_with_set_interval_includes()  {
+    for (set_interval1 <- test_intervals; set_interval2 <- test_intervals) {
+      val interval1 = set_interval1.interval
+      val interval2 = set_interval2.interval
+      assertEquals(interval1.includes(pord, interval2), set_interval1.includes(set_interval2))
+    }
+  }
+
   @Test def interval_agrees_with_set_interval_probably_overlaps() {
     for (set_interval1 <- test_intervals; set_interval2 <- test_intervals) {
       val interval1 = set_interval1.interval
@@ -226,6 +234,9 @@ case class SetInterval(start: Int, end: Int, includeStart: Boolean, includeEnd: 
   val interval: Interval = Interval(start, end, includeStart, includeEnd)
 
   def contains(point: Int): Boolean = doubledPointSet.contains(2 * point)
+
+  def includes(other: SetInterval): Boolean =
+    (other.doubledPointSet -- this.doubledPointSet).isEmpty
 
   def probablyOverlaps(other: SetInterval): Boolean = doubledPointSet.intersect(other.doubledPointSet).nonEmpty
 

--- a/src/test/scala/is/hail/utils/IntervalSuite.scala
+++ b/src/test/scala/is/hail/utils/IntervalSuite.scala
@@ -89,7 +89,7 @@ class IntervalSuite extends TestNGSuite {
     set_interval2 <- test_intervals} {
       val interval1 = set_interval1.interval
       val interval2 = set_interval2.interval
-      assertEquals(interval1.disjointAndGreaterThan(pord, interval2), set_interval1.disjointAndGreaterThan(set_interval2))
+      assertEquals(interval1.isAbove(pord, interval2), set_interval1.isAboveInterval(set_interval2))
     }
   }
 
@@ -98,7 +98,7 @@ class IntervalSuite extends TestNGSuite {
     set_interval2 <- test_intervals} {
       val interval1 = set_interval1.interval
       val interval2 = set_interval2.interval
-      assertEquals(interval1.disjointAndLessThan(pord, interval2), set_interval1.disjointAndLessThan(set_interval2))
+      assertEquals(interval1.isBelow(pord, interval2), set_interval1.isBelowInterval(set_interval2))
     }
   }
 
@@ -107,7 +107,7 @@ class IntervalSuite extends TestNGSuite {
     set_interval2 <- test_intervals} {
       val interval1 = set_interval1.interval
       val interval2 = set_interval2.interval
-      assertEquals(interval1.mergeable(pord, interval2), set_interval1.mergeable(set_interval2))
+      assertEquals(interval1.canMergeWith(pord, interval2), set_interval1.mergeable(set_interval2))
     }
   }
 
@@ -244,10 +244,10 @@ case class SetInterval(start: Int, end: Int, includeStart: Boolean, includeEnd: 
 
   def definitelyDisjoint(other: SetInterval): Boolean = doubledPointSet.intersect(other.doubledPointSet).isEmpty
 
-  def disjointAndGreaterThan(other: SetInterval): Boolean =
+  def isAboveInterval(other: SetInterval): Boolean =
     doubledPointSet.forall(p1 => other.doubledPointSet.forall(p2 => p1 > p2 ))
 
-  def disjointAndLessThan(other: SetInterval): Boolean =
+  def isBelowInterval(other: SetInterval): Boolean =
     doubledPointSet.forall(p1 => other.doubledPointSet.forall(p2 => p1 < p2 ))
 
   def mergeable(other: SetInterval): Boolean = {

--- a/src/test/scala/is/hail/utils/IntervalSuite.scala
+++ b/src/test/scala/is/hail/utils/IntervalSuite.scala
@@ -24,12 +24,12 @@ class IntervalSuite extends TestNGSuite {
 
   val test_itrees: IndexedSeq[SetIntervalTree] =
     SetIntervalTree(Array[(SetInterval, Int)]()) +:
-    test_intervals.flatMap { i1 =>
-      SetIntervalTree(Array(i1).zipWithIndex) +:
-      test_intervals.map { i2 =>
-        SetIntervalTree(Array(i1, i2).zipWithIndex)
-      }
-    } :+ SetIntervalTree(test_intervals.toArray.zipWithIndex)
+      test_intervals.flatMap { i1 =>
+        SetIntervalTree(Array(i1).zipWithIndex) +:
+          test_intervals.map { i2 =>
+            SetIntervalTree(Array(i1, i2).zipWithIndex)
+          }
+      } :+ SetIntervalTree(test_intervals.toArray.zipWithIndex)
 
 
   @Test def interval_agrees_with_set_interval_definitely_empty() {
@@ -42,14 +42,14 @@ class IntervalSuite extends TestNGSuite {
   @Test def interval_agrees_with_set_interval_greater_than_point() {
     for (set_interval <- test_intervals; p <- points) {
       val interval = set_interval.interval
-      assertEquals(interval.greaterThanPoint(pord, p), set_interval.doubledPointSet.forall(dp => dp > 2*p))
+      assertEquals(interval.isAbovePosition(pord, p), set_interval.doubledPointSet.forall(dp => dp > 2 * p))
     }
   }
 
   @Test def interval_agrees_with_set_interval_less_than_point() {
     for (set_interval <- test_intervals; p <- points) {
       val interval = set_interval.interval
-      assertEquals(interval.lessThanPoint(pord, p), set_interval.doubledPointSet.forall(dp => dp < 2*p))
+      assertEquals(interval.isBelowPosition(pord, p), set_interval.doubledPointSet.forall(dp => dp < 2 * p))
     }
   }
 
@@ -60,7 +60,7 @@ class IntervalSuite extends TestNGSuite {
     }
   }
 
-  @Test def interval_agrees_with_set_interval_includes()  {
+  @Test def interval_agrees_with_set_interval_includes() {
     for (set_interval1 <- test_intervals; set_interval2 <- test_intervals) {
       val interval1 = set_interval1.interval
       val interval2 = set_interval2.interval
@@ -245,10 +245,10 @@ case class SetInterval(start: Int, end: Int, includeStart: Boolean, includeEnd: 
   def definitelyDisjoint(other: SetInterval): Boolean = doubledPointSet.intersect(other.doubledPointSet).isEmpty
 
   def isAboveInterval(other: SetInterval): Boolean =
-    doubledPointSet.forall(p1 => other.doubledPointSet.forall(p2 => p1 > p2 ))
+    doubledPointSet.forall(p1 => other.doubledPointSet.forall(p2 => p1 > p2))
 
   def isBelowInterval(other: SetInterval): Boolean =
-    doubledPointSet.forall(p1 => other.doubledPointSet.forall(p2 => p1 < p2 ))
+    doubledPointSet.forall(p1 => other.doubledPointSet.forall(p2 => p1 < p2))
 
   def mergeable(other: SetInterval): Boolean = {
     val combinedPoints = doubledPointSet.union(other.doubledPointSet)
@@ -296,7 +296,7 @@ case class SetIntervalTree(annotations: Array[(SetInterval, Int)]) {
 
   val (intervals, values) = annotations.unzip
 
-  val annotationTree: IntervalTree[Int] = IntervalTree.annotationTree(pord, annotations.map { case (i, a) => (i.interval, a) } )
+  val annotationTree: IntervalTree[Int] = IntervalTree.annotationTree(pord, annotations.map { case (i, a) => (i.interval, a) })
 
   val intervalTree: IntervalTree[Unit] = IntervalTree(pord, intervals.map(_.interval))
 

--- a/src/test/scala/is/hail/utils/IntervalSuite.scala
+++ b/src/test/scala/is/hail/utils/IntervalSuite.scala
@@ -39,6 +39,20 @@ class IntervalSuite extends TestNGSuite {
     }
   }
 
+  @Test def interval_agrees_with_set_interval_greater_than_point() {
+    for (set_interval <- test_intervals; p <- points) {
+      val interval = set_interval.interval
+      assertEquals(interval.greaterThanPoint(pord, p), set_interval.doubledPointSet.forall(dp => dp > 2*p))
+    }
+  }
+
+  @Test def interval_agrees_with_set_interval_less_than_point() {
+    for (set_interval <- test_intervals; p <- points) {
+      val interval = set_interval.interval
+      assertEquals(interval.lessThanPoint(pord, p), set_interval.doubledPointSet.forall(dp => dp < 2*p))
+    }
+  }
+
   @Test def interval_agrees_with_set_interval_contains() {
     for (set_interval <- test_intervals; p <- points) {
       val interval = set_interval.interval

